src/query/graphite/native/aggregation_functions.go

```diff
@@ -278,6 +277,82 @@ func combineSeriesWithWildcards(
 	return r, nil
 }
 
+/*
+applyByNode takes a seriesList and applies some complicated function (described by a string), replacing templates with unique
+prefixes of keys from the seriesList (the key is all nodes up to the index given as `nodeNum`).
+
+If the `newName` parameter is provided, the name of the resulting series will be given by that parameter, with any
+"%" characters replaced by the unique prefix.
+
+Example:
+
+.. code-block:: none
+
+&target=applyByNode(servers.*.disk.bytes_free,1,"divideSeries(%.disk.bytes_free,sumSeries(%.disk.bytes_*))")
+
+Would find all series which match `servers.*.disk.bytes_free`, then trim them down to unique series up to the node
+given by nodeNum, then fill them into the template function provided (replacing % by the prefixes).
+
+Additional Examples:
+
+Given keys of
+
+- `stats.counts.haproxy.web.2XX`
+- `stats.counts.haproxy.web.3XX`
+- `stats.counts.haproxy.web.5XX`
+- `stats.counts.haproxy.microservice.2XX`
+- `stats.counts.haproxy.microservice.3XX`
+- `stats.counts.haproxy.microservice.5XX`
+
+The following will return the rate of 5XX's per service:
+
+.. code-block:: none
+
+applyByNode(stats.counts.haproxy.*.*XX, 3, "asPercent(%.5XX, sumSeries(%.*XX))", "%.pct_5XX")
+
+The output series would have keys `stats.counts.haproxy.web.pct_5XX` and `stats.counts.haproxy.microservice.pct_5XX`.
+*/
+func applyByNode(ctx *common.Context, seriesList singlePathSpec, nodeNum int, templateFunction string, newName string) (ts.SeriesList, error) {
+	// using this as a set
+	prefixMap := make(map[string]int)
+	for _, series := range seriesList.Values {
+		prefix := strings.Join(strings.Split(series.Name(), ".")[:nodeNum+1], ".")
```

Comment:
There may be an issue to using `series.Name()` here, since it's mutated when series are applied , e.g.
```
name := fmt.Sprintf("offsetToZero(%s)", series.Name())
series := ts.NewSeries(ctx, name, series.StartTime(), vals)
```
will change the name, so something like this may lead to using `offsetToZero(foo.bar` as the `prefix`; may need to add tracking for the original series name here? Unfortunately means it'll have to propagate everywhere we use `NewSeries`

Reply:
i think you're right but this is the way that they do it in the `graphite-web` repo:

```
prefixes = set()
for series in seriesList:
  prefix = '.'.join(series.name.split('.')[:nodeNum + 1])
  prefixes.add(prefix)
```

Reply:
ok it has been changed